elyra/util/archive.py

```diff
@@ -49,37 +69,58 @@ def tar_filter(tarinfo):
+            # which case we don't want to add this directory since recursive is False.
+            # This occurs with filenames like `data/util.py` or `data/*.py`.
+            elif not include_all and directory_in_list(tarinfo.name, filenames):
+                return tarinfo
+            return None
 
-        if '*' in files:
+        # We have a file at this point...
+
+        # Special case for single wildcard entries ('*')
+        if include_all:
             return tarinfo
 
-        for dependency in files:
-            if dependency:
-                if dependency.startswith('*'):
-                    # handle check for extension wildcard
-                    if tarinfo.name.endswith(dependency.replace('*', '')):
-                        return tarinfo
-                elif tarinfo.name == dependency:
-                    # handle check for specific file
-                    return tarinfo
-                elif recursive:
-                    # handle recursive
-                    return tarinfo
+        # Process filename
+        if filenames:
+            for filename in filenames:
+                if not filename or filename in processed_filenames:  # Skip processing
+                    continue
 
-        return None
+                # Match filename against candidate filename - handling wildcards
+                if fnmatch.fnmatch(tarinfo.name, filename):
+                    # if this is a direct match, record that its been processed
+                    if not has_wildcards(filename) and not recursive:
+                        processed_filenames.append(filename)
+                    matched_filenames.append(filename)
+                    return tarinfo
 
-    if files is None:
-        files = ['*']
+                # If the filename is a "flat" wildcarded value (i.e., isn't prefixed with a directory name)
+                # then we should take the basename of the candidate file to perform the match against.  This
+                # occurs for dependencies like *.py when include-subdirectories is enabled.
+                if not directory_prefixed(filename) and has_wildcards(filename):
+                    if fnmatch.fnmatch(os.path.basename(tarinfo.name), filename):
+                        matched_filenames.append(filename)
+                        return tarinfo
+            return None
 
+    # If there's a '*' - less things to check.
+    include_all = filenames and len(set([WILDCARDS[0]]) & set(filenames)) > 0
+    processed_filenames = []
+    matched_filenames = []
     temp_dir = create_project_temp_dir()
     archive = os.path.join(temp_dir, archive_name)
 
     with tarfile.open(archive, "w:gz") as tar:
         tar.add(source_dir, arcname="", filter=tar_filter)
 
-    if not archive:
-        raise RuntimeError('Internal error creating archive: {}'.format(archive_name))
+    if require_complete and not include_all:
+        # convert filenames and matched_filenames to sets and ensure they're the same.
```

Comment:
why not creating sets for both from the start instead of initiating them as arrays?

Reply:
Good question @karlaspuldaro.
Sets are immutable, so because `matched_filenames` gets built up during the processing, it can't be a `set` until we're done. `filenames`, on the other hand, comes into the method as a `list`, but I don't see it getting altered.  If python `sets` behave like `lists` in terms of iterables, then, yeah, `filenames` should be converted to a `set` first thing.
The other thing converting a `list` to a `set` does is remove duplicates - which I can't think of any reason why we wouldn't also want that.  Good idea.  I'll look at converting `filenames` to a `set` right away.  Thank you.

Reply:
Awesome, I've pushed the update.  It wound up simplifying indentation as well.  Thank you for the suggestion!
I'm going to mark this conversation as resolved.

Reply:
Turns out I was mistaken - Python sets are mutable - only `frozenset` is immutable.  Thank you to @karlaspuldaro for digging deeper into things!  I've just pushed a change that eliminates the need for conversion - thanks again Karla!